uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js

```diff
@@ -84,11 +91,76 @@ define([
       return !attachment.attachment_id || vm.canManage;
     };
 
+    /**
+     * Gets called when the component is destroyed
+     */
+    vm.$onDestroy = function () {
+      // destroy all the event
+      events.map(function (event) {
+        event();
+      });
+    };
+
     (function init () {
-      vm.supportedFileTypes = _.keys(sharedSettings.fileUploader.allowedMimeTypes);
-      vm.request.loadAttachments();
+      $q.all([
+        loadSupportedFileTypes(),
+        vm.request.loadAttachments()
+      ])
+      .then(function () {
+        $rootScope.$broadcast('Request Updated');
+      })
+      .finally(function () {
+        vm.filesLoaded = true;
+      });
+      events.push($rootScope.$on('uploadFiles: start', uploadAttachments));
     }());
 
+    /**
+     * Load file extensions which are supported for upload and creates uploader object
+     *
+     * @return {Promise}
+     */
+    function loadSupportedFileTypes () {
+      var allowedMimeTypes;
+
+      return OptionGroup.valuesOf('safe_file_extension')
+        .then(function (extensions) {
+          allowedMimeTypes = {};
+
+          vm.supportedFileTypes = extensions.map(function (ext) {
+            allowedMimeTypes[ext.label] = FileMimeTypes.getMimeTypeFor(ext.label);
+            return ext.label;
+          });
+        })
+        .finally(function () {
+          // if the API calls throws an error or fails, "allowedMimeTypes" will be undefined
+          // hence the default extension will be set to the uploader in file-upload.js
+          vm.fileUploader = FileUpload.uploader({
+            entityTable: 'civicrm_hrleaveandabsences_leave_request',
+            crmAttachmentToken: sharedSettings.attachmentToken,
+            queueLimit: sharedSettings.fileUploader.queueLimit,
+            allowedMimeTypes: allowedMimeTypes
+          });
+        });
+    }
+
+    /**
+     * Upload attachment in file uploder's queue, fires an event when done
+     */
+    function uploadAttachments () {
+      if (vm.fileUploader.queue && vm.fileUploader.queue.length > 0) {
+        vm.fileUploader.uploadAll({ entityID: vm.request.id })
+        .then(function () {
+          $rootScope.$broadcast('uploadFiles: success');
+        })
+        .catch(function () {
+          $rootScope.$broadcast('uploadFiles: error');
+        });
+      } else {
+        $rootScope.$broadcast('uploadFiles: success');
```

Comment:
the logic here is weird, why are we sending a success signal even if nothing happened?